pkg/cloud/services/compute/instance_test.go

```diff
@@ -366,3 +376,477 @@ func TestService_getServerNetworks(t *testing.T) {
+				m.ListImages(images.ListOpts{Name: "test-image"}).Return(
+					nil,
+					fmt.Errorf("test error"))
+			},
+			want:    "",
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.testName, func(t *testing.T) {
+			mockCtrl := gomock.NewController(t)
+			mockComputeClient := NewMockClient(mockCtrl)
+			tt.expect(mockComputeClient.EXPECT())
+
+			s := Service{
+				projectID:         "",
+				computeService:    mockComputeClient,
+				networkingService: &networking.Service{},
+				logger:            logr.Discard(),
+			}
+
+			got, err := s.getImageID(tt.imageName)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("Service.getImageID() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if got != tt.want {
+				t.Errorf("Service.getImageID() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+const (
+	networkUUID                   = "d412171b-9fd7-41c1-95a6-c24e5953974d"
+	subnetUUID                    = "d2d8d98d-b234-477e-a547-868b7cb5d6a5"
+	portUUID                      = "e7b7f3d1-0a81-40b1-bfa6-a22a31b17816"
+	imageUUID                     = "652b5a05-27fa-41d4-ac82-3e63cf6f7ab7"
+	flavorUUID                    = "6dc820db-f912-454e-a1e3-1081f3b8cc72"
+	instanceUUID                  = "383a8ec1-b6ea-4493-99dd-fc790da04ba9"
+	extraSecurityGroupUUID        = "514bb2d8-3390-4a3b-86a7-7864ba57b329"
+	controlPlaneSecurityGroupUUID = "c9817a91-4821-42db-8367-2301002ab659"
+	workerSecurityGroupUUID       = "9c6c0d28-03c9-436c-815d-58440ac2c1c8"
+	serverGroupUUID               = "7b940d62-68ef-4e42-a76a-1a62e290509c"
+
+	openStackMachineName = "test-openstack-machine"
+	portName             = "test-openstack-machine-0"
+	namespace            = "test-namespace"
+	imageName            = "test-image"
+	flavorName           = "test-flavor"
+	sshKeyName           = "test-ssh-key"
+)
+
+var failureDomain = "test-failure-domain"
+
+func getDefaultOpenStackCluster() *infrav1.OpenStackCluster {
+	return &infrav1.OpenStackCluster{
+		Spec: infrav1.OpenStackClusterSpec{},
+		Status: infrav1.OpenStackClusterStatus{
+			Network: &infrav1.Network{
+				ID: networkUUID,
+				Subnet: &infrav1.Subnet{
+					ID: subnetUUID,
+				},
+			},
+			ControlPlaneSecurityGroup: &infrav1.SecurityGroup{ID: controlPlaneSecurityGroupUUID},
+			WorkerSecurityGroup:       &infrav1.SecurityGroup{ID: workerSecurityGroupUUID},
+		},
+	}
+}
+
+func getDefaultMachine() *clusterv1.Machine {
+	return &clusterv1.Machine{
+		Spec: clusterv1.MachineSpec{
+			FailureDomain: &failureDomain,
+		},
+	}
+}
+
+func getDefaultOpenStackMachine() *infrav1.OpenStackMachine {
+	return &infrav1.OpenStackMachine{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      openStackMachineName,
+			Namespace: namespace,
+		},
+		Spec: infrav1.OpenStackMachineSpec{
+			// ProviderID is set by the controller
+			// InstanceID is set by the controller
+			// FloatingIP is only used by the cluster controller for the Bastion
+			// TODO: Test Networks, Ports, Subnet, and Trunk separately
```

Comment:
Mostly we have the networkingService covered by tests in the networking package now, I guess this is more to add some kind of integration test for both packages.

Reply:
Honestly I don't want to get too philosophical on what's a unit test and what's an integration test. I've been burned before by the maintenance burden and general uselessness of puritanically-scoped unit tests, so I've tried to keep it simple here.

* The scope of the test is the exported method (CreateInstance/DeleteInstance).
* The test boundary is what we can reasonably mock, which is currently the gophercloud calls.

If we were able to mock network service that might also be a practical boundary, but to be honest I quite the emergent benefit of revealing all the external API calls we're making, and they haven't felt *too* onerous yet.

Reply:
Yes, I do like that the tests show up all the API calls to OpenStack.